kernel/src/hil/public_key_crypto/keys.rs

```diff
@@ -283,3 +283,110 @@ pub trait RsaPrivKeyMut: PubPrivKeyMut + RsaKeyMut {
+/// is no assumption for implementers of this interface that keys can be added
+/// or changed or that keys can be specified by their actual bytes in a slice.
+///
+/// Selecting a key is asynchronous as it may require communication over a bus
+/// or waiting for an interrupt.
+///
+/// Keys are specified by an index starting at zero and going to
+/// `get_key_count()-1`.
+///
+/// The stored keys can be public or private keys.
+pub trait SelectKey<'a> {
+    /// Return the number of keys that the device can switch among.
+    ///
+    /// Each key must be identifiable by a consistent index.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `get_key_count_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if getting the count has started. Otherwise:
+    /// - `Err(ErrorCode::FAIL)` if the key count could not be started and there
+    ///   will be no callback.
+    fn get_key_count(&self) -> Result<(), ErrorCode>;
+
+    /// Set the key identified by its index as the active key.
+    ///
+    /// Indices start at 0 and go to `get_key_count() - 1`.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `select_key_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if the key select operation was accepted. Otherwise:
+    /// - `Err(ErrorCode::INVAL)` if the index is not valid.
+    /// - `Err(ErrorCode::ALREADY)` if the key is already selected.
+    fn select_key(&self, index: usize) -> Result<(), ErrorCode>;
+
+    fn set_client(&self, client: &'a dyn SelectKeyClient);
+}
+
+/// Client for setting keys.
+pub trait SetKeyClient<const KL: usize> {
+    /// Called when the key has been set.
+    ///
+    /// Returns the previous key if one was set.
+    ///
+    /// ### `error`:
+    ///
+    /// - `Ok(())`: The key was set successfully.
+    /// - `Err(())`: The key was not set successfully.
+    ///   - `ErrorCode::FAIL`: The key could not be set.
+    fn set_key_done(
+        &self,
+        previous_key: Option<&'static mut [u8; KL]>,
+        error: Result<(), ErrorCode>,
+    );
+}
+
+/// Interface for setting keys by a slice.
+///
+/// `KL` is the length of the keys.
+///
+/// Implementers must be able to store keys from a slice. This is most commonly
+/// used for implementations that hold keys in memory. However, this interface
+/// is asynchronous as keys may be stored in external storage or an external
+/// chip and require an asynchronous operations.
+pub trait SetKey<'a, const KL: usize> {
+    /// Set the current key.
+    ///
+    /// This is asynchronous. If there is an existing key that key will be
+    /// returned in the `set_key_done()` callback.
+    ///
+    /// ### Return
+    ///
+    /// `Ok()` if the key setting operation was accepted. Otherwise:
+    /// - `Err(ErrorCode::FAIL)` if the key cannot be set.
+    fn set_key(&self, key: &'static mut [u8; KL])
+        -> Result<(), (ErrorCode, &'static mut [u8; KL])>;
```

Comment:
This is probably regarding Tock's pub key interfaces more broadly, but if I had a secure element that was able to handle different types of keys, how would I implement that?

I can implement this trait multiple types, for different key lengths. But, for instance, in the case of Ed25519 and Curve P-256, they both use 256 bit keys. How would an implementation distinguish between those keys?

One possible solution could be for the implementation to provide wrapper types (e.g., a `struct MySecureElement` could be wrapped in a `struct MyEd25519SecureElement(MySecureElement)` and then implement `SetKey<'_, 32>`, thus allowing it to have different `SetKey<'_, 32>` implementations per key-type). But that seems less than elegant and usable.

Reply:
I don't know. The entire crypto stack is based on tradeoffs like these. I don't think we have a solution for a completely generic interface.